### PR TITLE
Group symex options into symex_config structure

### PR DESCRIPTION
**Pull request by romainbrenguier**
This makes it easy to separate the constant parameters of that class
from the other fields.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [na] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 936fd9c0c0c48b868e8103f28681950bae74850e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91653597

File: src/goto-symex/goto_symex.h [hunk 1]
```diff
@@ -38,6 +38,35 @@ class namespacet;
 class side_effect_exprt;
 class typecast_exprt;
 
+/// Configuration of the symbolic execution
+struct symex_configt
```
Comment (LAJW) [inline src/goto-symex/goto_symex.h]: ⛏ You might've forgotten `final`.

File: src/goto-symex/goto_symex.h [hunk 2]
```diff
@@ -38,6 +38,35 @@ class namespacet;
...
+  const bool doing_path_exploration;
+  const bool allow_pointer_unsoundness;
+  bool constant_propagation;
+  bool self_loops_to_assumptions;
+  bool simplify_opt;
+  bool unwinding_assertions;
+  bool partial_loops;
+  std::string debug_level;
+
+  explicit symex_configt(const optionst &options)
```
Comment (LAJW) [inline src/goto-symex/goto_symex.h]: 🤔 I can never decide whether parsing (or doing any operation that can fail) in the constructor is a good or bad idea. I wonder what happens when an option is not specified.

**Review by LAJW — APPROVED** [review]
The more typed options, the better.

File: src/goto-symex/goto_symex.h (hunk 2 above)
Comment (romainbrenguier) [reply src/goto-symex/goto_symex.h]: I could have a non-member `symex_configt symex_config(const optionst &)` function instead. Will try that.

Comment (LAJW) [reply src/goto-symex/goto_symex.h]: Making it a separate function only makes sense if the function can return an empty optional as an error for instance (constructors cannot return empty optionals).

Comment (romainbrenguier) [reply src/goto-symex/goto_symex.h]: The result of that is required for the goto_symext constructor anyway so using optional just there wouldn't solve the problem. I would suggest making the `goto_symext` constructor take a `symex_configt` instead.

Comment (romainbrenguier) [reply src/goto-symex/goto_symex.h]: Actually get_option functions have default for when the option is not specified (0 for int, "" for string) so we can go back to using a constructor and there is no reason for it to fail.

Comment (LAJW) [reply src/goto-symex/goto_symex.h]: Not a fan of silent failures, but this seems to be a subject for another time.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: bf2448294f479dd3f7026b8a83096760e48202da).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91678677

File: src/goto-symex/symex_builtin_functions.cpp [hunk 3]
```diff
@@ -458,7 +458,7 @@ void goto_symext::symex_trace(
 {
   PRECONDITION(code.arguments().size() >= 2);
 
-  int debug_thresh = unsafe_string2int(debug_level);
+  int debug_thresh = unsafe_string2int(symex_config.debug_level);
```
Comment (peterschrammel) [inline src/goto-symex/symex_builtin_functions.cpp]: Why not parse that right away?

**Comment by peterschrammel** [comment]
The benefit of this is not obvious, but I'm not strictly opposed.

**Comment by romainbrenguier** [comment]
@peterschrammel 
> The benefit of this is not obvious, but I'm not strictly opposed.

The benefits I see are that it simplifies the goto_symext class; it makes obvious in the code that a field comes from the configuration and thus is constant (they will all start with `symex_config.`); it makes clearer what are the options that can be passed to symex through the command line. We could also think of having an extra constructor using `symex_configt` instead of `optiont` and it will then be harder to forget about an option (not done in this PR so this argument doesn't really count here).

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7de21060a053729533e6cc28b930d84d35ae2121).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91842803

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 1dec39c0a889e98d6b6b06b275f44f73b2b97689).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91894990

**Review by tautschnig — APPROVED** [review]
Seems like a very good idea. Just a nit pick: there is a typo in the commit message: s/Pares/Parse/

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: bf8e58b6930641bc4998dd5205b96e7411f3f32e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/92545158

**Pull request merged**